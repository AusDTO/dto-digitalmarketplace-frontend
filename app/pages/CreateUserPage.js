import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { CreateUserForm } from '../components/CreateUser/CreateUserForm'
import { loadSignup, createUser } from '../actions/memberActions'
import ErrorBox from '../components/shared/form/ErrorBox'
import BaseForm from '../components/shared/form/BaseForm'
import formProps from '../components/shared/form/formPropsSelector'
import { withRouter } from 'react-router-dom'
import { rootPath } from '../routes'

export class CreateUserPage extends BaseForm {
  constructor(props) {
    super(props)
    this.state = {
      submitClicked: null,
      currentlySending: null
    }
  }

  componentWillMount() {
    const tokenString = this.props.location.pathname.substring(
      this.props.match.url.length + 1,
      this.props.location.pathname.length
    )
    this.props.loadRegistrationData(tokenString)
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.createUserSuccess !== nextProps.createUserSuccess) {
      if (nextProps.createUserSuccess) {
        this.props.history.push(`${rootPath}/signup/success/${nextProps.createUserForm.user_type}`)
      }
    }
  }

  onSubmitClicked = () => {
    this.setState({
      submitClicked: new Date().valueOf()
    })
  }

  render() {
    let {
      model,
      loadSignupSuccess,
      userRegisterDetails,
      handleSubmit,
      createUserSuccess,
      currentlySending
    } = this.props

    let hasFocused = false
    const setFocus = e => {
      if (!hasFocused) {
        hasFocused = true
        e.focus()
      }
    }

    return (
      <div className="row">
        <div className="col-sm-push-2 col-sm-8 col-xs-12">
          <article role="main">
            {loadSignupSuccess
              ? <CreateUserForm
                  initialState={userRegisterDetails}
                  onSubmitClicked={this.onSubmitClicked}
                  submitClicked={this.state.submitClicked}
                  handleSubmit={handleSubmit}
                  createUserSuccess={createUserSuccess}
                  model={model}
                  currentlySending={currentlySending}
                />
              : <ErrorBox title="There was a problem loading your details" setFocus={setFocus} />}
          </article>
        </div>
      </div>
    )
  }
}

CreateUserPage.propTypes = {
  userRegisterDetails: PropTypes.shape({
    name: PropTypes.string.isRequired,
    email_address: PropTypes.string.isRequired,
    manager_name: PropTypes.string,
    manager_email: PropTypes.string
  }),
  loadSignupSuccess: PropTypes.bool,
  loadRegistrationData: PropTypes.func.isRequired
}

const mapStateToProps = state => {
  return {
    ...formProps(state, 'createUserForm'),
    userRegisterDetails: state.user.userRegisterDetails,
    loadSignupSuccess: state.user.loadSignupSuccess,
    createUserSuccess: state.user.createUserSuccess,
    currentlySending: state.app.currentlySending
  }
}

const mapDispatchToProps = dispatch => {
  return {
    loadRegistrationData: tokenString => {
      dispatch(loadSignup(tokenString))
    },
    handleSubmit: values => dispatch(createUser(values))
  }
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(CreateUserPage))
