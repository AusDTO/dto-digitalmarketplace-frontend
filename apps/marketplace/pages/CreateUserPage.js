import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { parse } from 'qs'
import BaseForm from 'shared/form/BaseForm'
import formProps from 'shared/form/formPropsSelector'
import CreateUserForm from '../components/CreateUser/CreateUserForm'
import { createUser } from '../actions/memberActions'
import { rootPath } from '../routes'

export class CreateUserPageComponent extends BaseForm {
  constructor(props) {
    super(props)
    this.state = {
      submitClicked: null
    }

    this.onSubmitClicked = this.onSubmitClicked.bind(this)
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.createUserSuccess !== nextProps.createUserSuccess) {
      if (nextProps.createUserSuccess && this.props.createdUser) {
        this.props.history.push(`${rootPath}/signup/success/${nextProps.createdUser.role}`)
      }
    }
  }

  onSubmitClicked() {
    this.setState({
      submitClicked: new Date().valueOf()
    })
  }

  getTokenFromURL() {
    const tokenString = this.props.location.pathname.substring(
      this.props.match.url.length + 1,
      this.props.location.pathname.length
    )
    return tokenString
  }

  getEmailFromQueryString() {
    const parsed = parse(this.props.location.search.substr(1))
    let emailAddress = ''
    if (parsed.e) {
      emailAddress = parsed.e
    }
    return emailAddress
  }

  handleSubmit = values => {
    const token = this.getTokenFromURL()
    const email = this.getEmailFromQueryString()
    const password = values.password
    this.props.createUser(token, email, password)
  }

  render() {
    const {
      model,
      createUserSuccess,
      currentlySending
    } = this.props

    return (
      <div className="row">
        <div className="col-sm-push-2 col-sm-8 col-xs-12">
          <article role="main">
            <CreateUserForm
              onSubmitClicked={this.onSubmitClicked}
              submitClicked={this.state.submitClicked}
              handleSubmit={this.handleSubmit}
              createUserSuccess={createUserSuccess}
              model={model}
              currentlySending={currentlySending}
            />
          </article>
        </div>
      </div>
    )
  }
}

CreateUserPageComponent.propTypes = {
  model: PropTypes.string.isRequired,
  createUserSuccess: PropTypes.bool
}

const mapStateToProps = state => ({
  ...formProps(state, 'createUserForm'),
  createUserSuccess: state.user.createUserSuccess,
  createdUser: state.user.user,
  currentlySending: state.app.currentlySending
})

const mapDispatchToProps = dispatch => ({
  createUser: (tokenString, emailAddress, password) =>
    dispatch(createUser(tokenString, emailAddress, password))
})

const CreateUserPage = withRouter(connect(mapStateToProps, mapDispatchToProps)(CreateUserPageComponent))

export default CreateUserPage
