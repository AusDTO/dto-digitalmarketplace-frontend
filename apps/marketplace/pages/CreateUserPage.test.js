import React from 'react'
import renderer from 'react-test-renderer'
import expect from 'expect'
import { MemoryRouter } from 'react-router-dom'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import configureStore from 'marketplace/store'
import { Provider } from 'react-redux'

import CreateUserPage from './CreateUserPage'

Enzyme.configure({ adapter: new Adapter() })

jest.mock('shared/Icon/_getIcons')

describe('Test suite for CreateUserPage page', () => {
  const state = {
    userRegistrationDetails: {
      name: 'Jeff Labowski',
      email_address: 'e@mail.com'
    },
    loadCompleteSuccess: true,
    loadRegistrationData: () => true
  }

  it('CreateUserPage renders without errors', () => {
    const store = configureStore(state)

    const wrapper = shallow(
      <Provider store={store}>
        <MemoryRouter>
          <CreateUserPage {...state} />
        </MemoryRouter>
      </Provider>
    )
    expect(wrapper).toBeDefined()
  })

  it('Because snaphots', () => {
    const store = configureStore(state)
    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <CreateUserPage {...state} />
        </MemoryRouter>
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
