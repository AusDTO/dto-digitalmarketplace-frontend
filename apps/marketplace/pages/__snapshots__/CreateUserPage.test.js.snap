// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suite for CreateUserPage page Because snaphots 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-push-2 col-sm-8 col-xs-12"
  >
    <article
      role="main"
    >
      <div>
        <div>
          <h1
            className="au-display-xl"
          >
            Add a password
          </h1>
          <p>
            To finish creating your account please provide the following details.
          </p>
        </div>
        <form
          id="createuser"
          onReset={[Function]}
          onSubmit={[Function]}
        >
          <div
            className="field "
          >
            <label
              className="question-heading au-text-input__label"
              htmlFor="password"
              id="password-label"
            >
              Password
            </label>
            <div
              className="au-text-input__hint"
              id="password-hint"
            >
              At least 10 characters
            </div>
            <div
              className="table"
            >
              <input
                aria-describedby="password-hint"
                className="au-text-input au-text-input--block"
                defaultValue=""
                disabled={false}
                id="password"
                max={null}
                maxLength={0}
                min={null}
                minLength={0}
                name="password"
                onChange={[Function]}
                pattern={null}
                placeholder=""
                readOnly={false}
                type="password"
                value=""
              />
            </div>
          </div>
          <label
            className="au-control-input"
            htmlFor="agree"
          >
            <input
              checked={false}
              className="au-control-input__input"
              disabled={false}
              id="agree"
              name="agree"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyPress={[Function]}
              type="checkbox"
              value="agree"
            />
            <span
              className="au-control-input__text"
            >
              <span>
                <span>
                  I accept the 
                </span>
                <a
                  href="/terms-of-use"
                  rel="external"
                  target="_blank"
                >
                  Terms of Use
                </a>
              </span>
            </span>
            
          </label>
          <div
            className="formSubmitBtnWrapper"
          >
            <input
              className="au-btn"
              onClick={[Function]}
              type="submit"
              value="Join the Marketplace"
            />
          </div>
        </form>
      </div>
    </article>
  </div>
</div>
`;
